pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,352 @@
+		values := make([]uint32, 1)
+
+		// Do a batch update, check the result.
+		// We do an update instead of a lookup because it's more reliable for detection
+		_, err = m.BatchUpdate(keys, values, nil)
+		batchAPISupported = err == nil
+	})
+	return batchAPISupported
+}
+
+// GenericMap is a wrapper around ebpf.Map that allows to use generic types.
+// Also includes support for batch iterations
+type GenericMap[K interface{}, V interface{}] struct {
+	m *ebpf.Map
+}
+
+// NewGenericMap creates a new GenericMap with the given spec. Key and Value sizes are automatically
+// inferred from the types of K and V.
+// Important: if the map is a per-cpu map, V must be a slice type
+func NewGenericMap[K interface{}, V interface{}](spec *ebpf.MapSpec) (*GenericMap[K, V], error) {
+	// Automatic inference of sizes. We assume that K/V are simple types that
+	// can be instantiated with no arguments
+	var kval K
+	var vval V
+
+	err := validateValueTypeForMapType[V](spec.Type)
+	if err != nil {
+		return nil, err
+	}
+
+	spec.KeySize = uint32(unsafe.Sizeof(kval))
+
+	if isPerCPU(spec.Type) {
+		spec.ValueSize = uint32(reflect.TypeOf(vval).Elem().Size())
+	} else {
+		spec.ValueSize = uint32(unsafe.Sizeof(vval))
+	}
+
+	m, err := ebpf.NewMap(spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return &GenericMap[K, V]{
+		m: m,
+	}, nil
+}
+
+// validateValueTypeForMapType checks that the type of values (V) is valid for the given map type, returning
+// an error if it's not valid.
+//
+// For now it ensures that per-cpu maps use a slice type for the value.
+// Separate function to allow using it in the different constructors/converters
+func validateValueTypeForMapType[V interface{}](t ebpf.MapType) error {
+	var vval V
+	if isPerCPU(t) && reflect.TypeOf(vval).Kind() != reflect.Slice {
+		return fmt.Errorf("per-cpu maps require a slice type for the value, instead got %T", vval)
+	}
+	return nil
+}
+
+// Map creates a new GenericMap from an existing ebpf.Map
+func Map[K interface{}, V interface{}](m *ebpf.Map) (*GenericMap[K, V], error) {
+	if err := validateValueTypeForMapType[V](m.Type()); err != nil {
+		return nil, err
+	}
+
+	return &GenericMap[K, V]{
+		m: m,
+	}, nil
+}
+
+// GetMap gets the generic map with the given name from the manager
+func GetMap[K interface{}, V interface{}](mgr *manager.Manager, name string) (*GenericMap[K, V], error) {
+	m, _, err := mgr.GetMap(name)
+	if err != nil {
+		return nil, err
+	}
+	if m == nil {
+		return nil, fmt.Errorf("map %q not found", name)
+	}
+	gm, err := Map[K, V](m)
+	if err != nil {
+		return nil, err
+	}
+	return gm, nil
+}
+
+// Map returns the underlying ebpf.Map
+func (g *GenericMap[K, V]) Map() *ebpf.Map {
+	return g.m
+}
+
+// IteratorOptions are options for the Iterate method
+type IteratorOptions struct {
+	BatchSize       int  // Number of items to fetch per batch. If 0, use default value (100)
+	ForceSingleItem bool // Force the use of the single item iterator even if the batch API is supported
```

Comment:
Could this be handled by setting `BatchSize = 1`?

Reply:
We could do that, in that case we could even get rid of the IteratorOptions altogether.